File: backend/routes/integrations.py
from fastapi import APIRouter, HTTPException, status, Depends
from typing import List, Optional, Dict, Any
import httpx
import os
from ..models.user import User
from ..auth.dependencies import get_current_user, require_admin_or_manager
from ..database.mongodb import DatabaseOperations
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/integrations", tags=["integrations"])

# Integration Models
class SlackIntegration:
    def __init__(self, webhook_url: str):
        self.webhook_url = webhook_url
    
    async def send_notification(self, message: str, channel: str = None):
        """Send notification to Slack"""
        try:
            payload = {
                "text": message,
                "username": "Hubstaff Bot"
            }
            if channel:
                payload["channel"] = channel
            
            async with httpx.AsyncClient() as client:
                response = await client.post(self.webhook_url, json=payload)
                return response.status_code == 200
        except Exception as e:
            logger.error(f"Slack notification error: {e}")
            return False

class TrelloIntegration:
    def __init__(self, api_key: str, token: str):
        self.api_key = api_key
        self.token = token
        self.base_url = "https://api.trello.com/1"
    
    async def create_card(self, list_id: str, name: str, description: str = None):
        """Create a Trello card"""
        try:
            url = f"{self.base_url}/cards"
            params = {
                "key": self.api_key,
                "token": self.token,
                "idList": list_id,
                "name": name
            }
            if description:
                params["desc"] = description
            
            async with httpx.AsyncClient() as client:
                response = await client.post(url, params=params)
                return response.json() if response.status_code == 200 else None
        except Exception as e:
            logger.error(f"Trello card creation error: {e}")
            return None
    
    async def get_boards(self):
        """Get user's Trello boards"""
        try:
            url = f"{self.base_url}/members/me/boards"
            params = {
                "key": self.api_key,
                "token": self.token
            }
            
            async with httpx.AsyncClient() as client:
                response = await client.get(url, params=params)
                return response.json() if response.status_code == 200 else []
        except Exception as e:
            logger.error(f"Trello boards fetch error: {e}")
            return []

class GitHubIntegration:
    def __init__(self, token: str):
        self.token = token
        self.base_url = "https://api.github.com"
        self.headers = {
            "Authorization": f"token {token}",
            "Accept": "application/vnd.github.v3+json"
        }
    
    async def create_issue(self, repo: str, title: str, body: str = None, labels: List[str] = None):
        """Create a GitHub issue"""
        try:
            url = f"{self.base_url}/repos/{repo}/issues"
            payload = {
                "title": title,
                "body": body or "",
                "labels": labels or []
            }
            
            async with httpx.AsyncClient() as client:
                response = await client.post(url, json=payload, headers=self.headers)
                return response.json() if response.status_code == 201 else None
        except Exception as e:
            logger.error(f"GitHub issue creation error: {e}")
            return None
    
    async def get_repositories(self):
        """Get user's GitHub repositories"""
        try:
            url = f"{self.base_url}/user/repos"
            
            async with httpx.AsyncClient() as client:
                response = await client.get(url, headers=self.headers)
                return response.json() if response.status_code == 200 else []
        except Exception as e:
            logger.error(f"GitHub repos fetch error: {e}")
            return []

# Integration Routes
@router.post("/slack/connect")
async def connect_slack(
    webhook_url: str,
    current_user: User = Depends(get_current_user)
):
    """Connect Slack integration"""
    try:
        # Test the webhook
        slack = SlackIntegration(webhook_url)
        success = await slack.send_notification("Hubstaff integration connected successfully!")
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid Slack webhook URL"
            )
        
        # Store integration
        integration_data = {
            "user_id": current_user.id,
            "type": "slack",
            "config": {"webhook_url": webhook_url},
            "active": True,
            "created_at": "datetime.utcnow()"
        }
        
        await DatabaseOperations.create_document("integrations", integration_data)
        
        return {"message": "Slack integration connected successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Slack connection error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to connect Slack integration"
        )

@router.post("/trello/connect")
async def connect_trello(
    api_key: str,
    token: str,
    current_user: User = Depends(get_current_user)
):
    """Connect Trello integration"""
    try:
        # Test the credentials
        trello = TrelloIntegration(api_key, token)
        boards = await trello.get_boards()
        
        if not boards:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid Trello credentials"
            )
        
        # Store integration
        integration_data = {
            "user_id": current_user.id,
            "type": "trello",
            "config": {"api_key": api_key, "token": token},
            "active": True,
            "created_at": "datetime.utcnow()"
        }
        
        await DatabaseOperations.create_document("integrations", integration_data)
        
        return {
            "message": "Trello integration connected successfully",
            "boards": boards[:5]  # Return first 5 boards
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Trello connection error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to connect Trello integration"
        )

@router.post("/github/connect")
async def connect_github(
    token: str,
    current_user: User = Depends(get_current_user)
):
    """Connect GitHub integration"""
    try:
        # Test the token
        github = GitHubIntegration(token)
        repos = await github.get_repositories()
        
        if not repos:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Invalid GitHub token"
            )
        
        # Store integration
        integration_data = {
            "user_id": current_user.id,
            "type": "github",
            "config": {"token": token},
            "active": True,
            "created_at": "datetime.utcnow()"
        }
        
        await DatabaseOperations.create_document("integrations", integration_data)
        
        return {
            "message": "GitHub integration connected successfully",
            "repositories": repos[:10]  # Return first 10 repos
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"GitHub connection error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to connect GitHub integration"
        )

@router.get("/")
async def get_integrations(current_user: User = Depends(get_current_user)):
    """Get user's integrations"""
    try:
        integrations = await DatabaseOperations.get_documents(
            "integrations",
            {"user_id": current_user.id, "active": True}
        )
        
        # Remove sensitive data
        for integration in integrations:
            if "config" in integration:
                config = integration["config"]
                if "token" in config:
                    config["token"] = "***"
                if "api_key" in config:
                    config["api_key"] = "***"
        
        return {"integrations": integrations}
        
    except Exception as e:
        logger.error(f"Get integrations error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get integrations"
        )

@router.post("/slack/notify")
async def send_slack_notification(
    message: str,
    channel: str = None,
    current_user: User = Depends(get_current_user)
):
    """Send Slack notification"""
    try:
        # Get Slack integration
        integration = await DatabaseOperations.get_document(
            "integrations",
            {"user_id": current_user.id, "type": "slack", "active": True}
        )
        
        if not integration:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Slack integration not found"
            )
        
        webhook_url = integration["config"]["webhook_url"]
        slack = SlackIntegration(webhook_url)
        
        success = await slack.send_notification(message, channel)
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to send Slack notification"
            )
        
        return {"message": "Notification sent successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Slack notification error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to send notification"
        )

@router.post("/trello/create-card")
async def create_trello_card(
    list_id: str,
    name: str,
    description: str = None,
    current_user: User = Depends(get_current_user)
):
    """Create Trello card from task"""
    try:
        # Get Trello integration
        integration = await DatabaseOperations.get_document(
            "integrations",
            {"user_id": current_user.id, "type": "trello", "active": True}
        )
        
        if not integration:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Trello integration not found"
            )
        
        config = integration["config"]
        trello = TrelloIntegration(config["api_key"], config["token"])
        
        card = await trello.create_card(list_id, name, description)
        
        if not card:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to create Trello card"
            )
        
        return {"message": "Trello card created successfully", "card": card}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Trello card creation error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create Trello card"
        )

@router.post("/github/create-issue")
async def create_github_issue(
    repo: str,
    title: str,
    body: str = None,
    labels: List[str] = None,
    current_user: User = Depends(get_current_user)
):
    """Create GitHub issue from task"""
    try:
        # Get GitHub integration
        integration = await DatabaseOperations.get_document(
            "integrations",
            {"user_id": current_user.id, "type": "github", "active": True}
        )
        
        if not integration:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="GitHub integration not found"
            )
        
        token = integration["config"]["token"]
        github = GitHubIntegration(token)
        
        issue = await github.create_issue(repo, title, body, labels)
        
        if not issue:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to create GitHub issue"
            )
        
        return {"message": "GitHub issue created successfully", "issue": issue}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"GitHub issue creation error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create GitHub issue"
        )

@router.delete("/{integration_id}")
async def disconnect_integration(
    integration_id: str,
    current_user: User = Depends(get_current_user)
):
    """Disconnect integration"""
    try:
        integration = await DatabaseOperations.get_document(
            "integrations",
            {"id": integration_id, "user_id": current_user.id}
        )
        
        if not integration:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Integration not found"
            )
        
        await DatabaseOperations.update_document(
            "integrations",
            {"id": integration_id},
            {"active": False}
        )
        
        return {"message": "Integration disconnected successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Disconnect integration error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to disconnect integration"
        )

File: backend/routes/integrations.js
const express = require('express');
const { authMiddleware, requireManager } = require('../middleware/auth');

const router = express.Router();

// Get available integrations
router.get('/', authMiddleware, async (req, res) => {
  try {
    const integrations = [
      {
        id: 'slack',
        name: 'Slack',
        description: 'Connect with Slack for team notifications',
        icon: '💬',
        category: 'Communication',
        status: 'available',
        configured: false,
        features: [
          'Time tracking notifications',
          'Project updates',
          'Daily summaries',
          'Team alerts'
        ]
      },
      {
        id: 'trello',
        name: 'Trello',
        description: 'Sync tasks with Trello boards',
        icon: '📋',
        category: 'Project Management',
        status: 'available',
        configured: false,
        features: [
          'Task synchronization',
          'Board integration',
          'Card time tracking',
          'Progress updates'
        ]
      },
      {
        id: 'github',
        name: 'GitHub',
        description: 'Track time on GitHub issues and pull requests',
        icon: '🐙',
        category: 'Development',
        status: 'available',
        configured: false,
        features: [
          'Issue tracking',
          'Pull request time',
          'Commit tracking',
          'Repository insights'
        ]
      },
      {
        id: 'jira',
        name: 'Jira',
        description: 'Integrate with Jira for issue tracking',
        icon: '🎯',
        category: 'Project Management',
        status: 'available',
        configured: false,
        features: [
          'Issue synchronization',
          'Sprint tracking',
          'Worklog integration',
          'Status updates'
        ]
      },
      {
        id: 'calendar',
        name: 'Google Calendar',
        description: 'Sync with Google Calendar events',
        icon: '📅',
        category: 'Productivity',
        status: 'available',
        configured: false,
        features: [
          'Calendar sync',
          'Meeting tracking',
          'Event integration',
          'Schedule optimization'
        ]
      },
      {
        id: 'zapier',
        name: 'Zapier',
        description: 'Connect with 3000+ apps via Zapier',
        icon: '⚡',
        category: 'Automation',
        status: 'available',
        configured: false,
        features: [
          'Workflow automation',
          'Custom triggers',
          'Data synchronization',
          'Cross-platform integration'
        ]
      }
    ];
    
    res.json({
      integrations,
      categories: ['Communication', 'Project Management', 'Development', 'Productivity', 'Automation']
    });
  } catch (error) {
    console.error('Get integrations error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get integration details
router.get('/:id', authMiddleware, async (req, res) => {
  try {
    const { id } = req.params;
    
    // Mock integration details - in real implementation, this would come from database
    const integrationDetails = {
      slack: {
        id: 'slack',
        name: 'Slack',
        description: 'Connect with Slack for team notifications',
        icon: '💬',
        category: 'Communication',
        status: 'available',
        configured: false,
        setup_steps: [
          'Create a Slack app in your workspace',
          'Add bot permissions: chat:write, channels:read',
          'Install the app to your workspace',
          'Copy the Bot User OAuth Token',
          'Paste the token in the configuration below'
        ],
        config_fields: [
          {
            name: 'bot_token',
            label: 'Bot User OAuth Token',
            type: 'password',
            required: true,
            description: 'Your Slack bot token starting with xoxb-'
          },
          {
            name: 'channel',
            label: 'Default Channel',
            type: 'text',
            required: false,
            description: 'Channel for notifications (e.g., #general)'
          }
        ]
      },
      trello: {
        id: 'trello',
        name: 'Trello',
        description: 'Sync tasks with Trello boards',
        icon: '📋',
        category: 'Project Management',
        status: 'available',
        configured: false,
        setup_steps: [
          'Get your Trello API key from https://trello.com/app-key',
          'Generate a token for your application',
          'Find your board ID from the board URL',
          'Configure the integration below'
        ],
        config_fields: [
          {
            name: 'api_key',
            label: 'API Key',
            type: 'password',
            required: true
          },
          {
            name: 'token',
            label: 'Token',
            type: 'password',
            required: true
          },
          {
            name: 'board_id',
            label: 'Board ID',
            type: 'text',
            required: true
          }
        ]
      },
      github: {
        id: 'github',
        name: 'GitHub',
        description: 'Track time on GitHub issues and pull requests',
        icon: '🐙',
        category: 'Development',
        status: 'available',
        configured: false,
        setup_steps: [
          'Create a Personal Access Token in GitHub',
          'Grant repo permissions to the token',
          'Add the token to the configuration below',
          'Select repositories to track'
        ],
        config_fields: [
          {
            name: 'access_token',
            label: 'Personal Access Token',
            type: 'password',
            required: true
          },
          {
            name: 'repositories',
            label: 'Repositories',
            type: 'textarea',
            required: true,
            description: 'List repositories (one per line, format: owner/repo)'
          }
        ]
      }
    };
    
    const integration = integrationDetails[id];
    if (!integration) {
      return res.status(404).json({ error: 'Integration not found' });
    }
    
    res.json(integration);
  } catch (error) {
    console.error('Get integration details error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Configure integration (admin/manager only)
router.post('/:id/configure', authMiddleware, requireManager, async (req, res) => {
  try {
    const { id } = req.params;
    const { config } = req.body;
    
    // Mock configuration save - in real implementation, this would save to database
    // and validate the configuration
    
    res.json({
      message: `${id} integration configured successfully`,
      integration: {
        id,
        configured: true,
        config: {
          // Return sanitized config (without sensitive data)
          ...config,
          // Remove sensitive fields
          bot_token: config.bot_token ? '***' : undefined,
          api_key: config.api_key ? '***' : undefined,
          access_token: config.access_token ? '***' : undefined,
          token: config.token ? '***' : undefined
        }
      }
    });
  } catch (error) {
    console.error('Configure integration error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Test integration connection
router.post('/:id/test', authMiddleware, requireManager, async (req, res) => {
  try {
    const { id } = req.params;
    
    // Mock test - in real implementation, this would test the actual connection
    const testResults = {
      slack: {
        success: true,
        message: 'Successfully connected to Slack workspace',
        details: {
          workspace: 'My Team Workspace',
          bot_user: 'Hubstaff Bot',
          permissions: ['chat:write', 'channels:read']
        }
      },
      trello: {
        success: true,
        message: 'Successfully connected to Trello',
        details: {
          boards_found: 5,
          user: 'john.doe@example.com'
        }
      },
      github: {
        success: true,
        message: 'Successfully connected to GitHub',
        details: {
          user: 'johndoe',
          repositories: 12,
          permissions: ['repo']
        }
      }
    };
    
    const result = testResults[id] || {
      success: false,
      message: 'Integration test not implemented'
    };
    
    res.json(result);
  } catch (error) {
    console.error('Test integration error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Disable integration
router.post('/:id/disable', authMiddleware, requireManager, async (req, res) => {
  try {
    const { id } = req.params;
    
    // Mock disable - in real implementation, this would update database
    
    res.json({
      message: `${id} integration disabled successfully`,
      integration: {
        id,
        configured: false
      }
    });
  } catch (error) {
    console.error('Disable integration error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// Get integration logs
router.get('/:id/logs', authMiddleware, requireManager, async (req, res) => {
  try {
    const { id } = req.params;
    const { page = 1, limit = 20 } = req.query;
    
    // Mock logs - in real implementation, this would come from database
    const logs = [
      {
        id: '1',
        timestamp: new Date(),
        level: 'info',
        message: 'Integration sync completed successfully',
        details: {
          items_synced: 5,
          duration: '2.3s'
        }
      },
      {
        id: '2',
        timestamp: new Date(Date.now() - 3600000),
        level: 'warning',
        message: 'Rate limit reached, retrying in 60 seconds',
        details: {
          retry_count: 1
        }
      },
      {
        id: '3',
        timestamp: new Date(Date.now() - 7200000),
        level: 'error',
        message: 'Failed to sync project: Invalid project ID',
        details: {
          project_id: 'invalid-123',
          error_code: 'PROJECT_NOT_FOUND'
        }
      }
    ];
    
    res.json({
      logs,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: logs.length,
        pages: Math.ceil(logs.length / limit)
      }
    });
  } catch (error) {
    console.error('Get integration logs error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;